### PR TITLE
Microchip icicle kit platform support

### PR DESCRIPTION
**Pull request by codehub8**
+ HardenOS1.0
+ Address sanitizer disabled
+ By default docker is disabled

### CODE REVIEW

**Comment by vilvo** [comment]
Thanks a lot for your PR. We're in the process of enabling out of tree ghaf-configurations - which does not mean we could not merge this as RISC-V reference - so I'll share you the a recent outline for the ghaf structure development that has impact to the review of this PR. Please see the outline of coming ghaf changes below:
image

Given above, I'd like you to clarify if you want to introduce this target as new reference and maintain it.
If that's open, we could also consider this PR as candidate for out-of-tree ghaf configuration. There's recent template support for it available https://github.com/tiiuae/ghaf/pull/164 
Please let me know your view.

File: docs/src/features/features.md
Comment (vilvo) [inline docs/src/features/features.md:36]: Depending on `nixpkgs 22.11` in ghaf tree is problematic. What are the exact issues for `riscv64` support on `nixpkgs 23.05`? I've understood it has better cross-compilation support in general and the cross-compilation support will be improved also for ghaf dependencies on `aarch64`

File: docs/src/ref_impl/build_and_run.md
Comment (vilvo) [inline docs/src/ref_impl/build_and_run.md:35]: Please see my comment on a new reference configuration https://github.com/tiiuae/ghaf/pull/165#issuecomment-1581997751

File: docs/src/ref_impl/build_and_run.md
Comment (vilvo) [inline docs/src/ref_impl/build_and_run.md:175]: Should we also instruct here to use the `-debug` target? Considering other reference targets, `-release` is of very limited usefulness. Also, all the possible debug channels (ssh|serial|..) should go to `-debug`-targets.

File: flake.nix
Comment (vilvo) [inline flake.nix:20]: Nice work on commenting the details of cross-compilation issues this way! Excellent.

File: flake.nix
Comment (vilvo) [inline flake.nix:27]: Please link the pending PR here in the comments with `#TODO: https://github.com/NixOS/nixos-hardware/pull/612`

File: modules/hardware/polarfire/base-packages.nix
Comment (vilvo) [inline modules/hardware/polarfire/base-packages.nix:1]: This goes against ghaf reference configuration defined packages on host - that will be more minimal in the next phases. See https://tiiuae.github.io/ghaf/features/features.html#next-steps

As these are not essential to the configuration I'd propose two things:
- please refactor this to a development module
- make them user (`ghaf`) packages or packages for user passed passed as argument to the module. As `systemPackages` they are made available to all users (incl. `root`) which should be avoided.

File: modules/hardware/polarfire/configuration.nix
Comment (vilvo) [inline modules/hardware/polarfire/configuration.nix:59]: This should be used from common definitions for reference configuration - see https://github.com/tiiuae/ghaf/blob/44a8660de954404c5ce1b47d31f89136b941ec97/modules/host/default.nix#L19
Hardware specifics should not override / conflict / duplicate with the common `networking.hostName` - this will keep a ghaf system on different hardware more familiar for developers/users.

File: modules/hardware/polarfire/configuration.nix
Comment (vilvo) [inline modules/hardware/polarfire/configuration.nix:97]: Please see my earlier comment on ghaf docker support.

File: modules/hardware/polarfire/configuration.nix
Comment (vilvo) [inline modules/hardware/polarfire/configuration.nix:106]: Overlaps with https://github.com/tiiuae/ghaf/commit/a21683501fba2adbd71227343feb6d2473ce770a#diff-b97fa41267366eb5a1b917982f2ec07f9aa1ecac38caaa167d6027dc511adfb4R14

File: modules/hardware/polarfire/configuration.nix
Comment (vilvo) [inline modules/hardware/polarfire/configuration.nix:103]: Networking to `networking.nix` module

File: templates/targets/riscv64/microchip/polarfire/flake.nix
Comment (vilvo) [inline templates/targets/riscv64/microchip/polarfire/flake.nix:1]: Ok, this kind of addresses my earlier question on whether this should be host-only (no virtualisation) reference or out-of-tree ghaf configuration. This template aims to make this a general ghaf reference configuration target.

File: modules/hardware/polarfire/mpfs-nixos-sdimage.nix
Comment (vilvo) [inline modules/hardware/polarfire/mpfs-nixos-sdimage.nix:1]: Thanks, this should be tested by another person. We can come back to it when we've worked with the other structural changes I requested.

File: modules/hardware/polarfire/debug-packages.nix
Comment (vilvo) [inline modules/hardware/polarfire/debug-packages.nix:3]: Most of these come from stdenv and can be used in `callPackage` with whatever nix declared development packages. We should not put them to host and to `systemPackages` - who ever needs these can use specific `nix-shell -p <package>`or `nix develop` with flakes. Leaving them out will also reduce the image size and decrease build time.

File: modules/hardware/polarfire/mpfs-nixos-sdimage.nix
Comment (codehub8) [reply modules/hardware/polarfire/mpfs-nixos-sdimage.nix:1]: OK, I will push some commits with the structural changes.

**Review by vilvo — CHANGES_REQUESTED** [review]
Please address questions and/or structural change requests.

File: modules/hardware/polarfire/mpfs-nixos-sdimage.nix
Comment (vilvo) [reply modules/hardware/polarfire/mpfs-nixos-sdimage.nix:1]: Thanks, please squash (or `--amend`) the future commits in the branch into one and force push to your PR branch. 

Please also note that there was automated nix fmt-check failing and should be fixed.

Comment (vilvo) [reply modules/hardware/polarfire/mpfs-nixos-sdimage.nix:1]: And commits must be signed - see https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification
in addition to DCO `Signed-off-by:`
Branch protection will block unsigned commits from getting merged.

File: docs/src/features/features.md
Comment (codehub8) [reply docs/src/features/features.md:36]: There are few issues I found in cross compilation for RiscV platform.
Listed the same in code also.
[BUG:235179](https://github.com/NixOS/nixpkgs/issues/235179)
[BUG:235526](https://github.com/NixOS/nixpkgs/issues/235526)

File: docs/src/ref_impl/build_and_run.md
Comment (codehub8) [reply docs/src/ref_impl/build_and_run.md:35]: Discussed with other stakeholders and we decided to integrate Polarfire as reference platform for now.

File: docs/src/ref_impl/build_and_run.md
Comment (codehub8) [reply docs/src/ref_impl/build_and_run.md]: Added instructions to build image for debug target too.

File: flake.nix
Comment (codehub8) [reply flake.nix:20]: Thank you :)

File: flake.nix
Comment (codehub8) [reply flake.nix:31]: I have listed justification for a forked repository in the file.

File: modules/hardware/polarfire/configuration.nix
Comment (codehub8) [reply modules/hardware/polarfire/configuration.nix]: Removed this file, created new common file for headless platforms.

File: modules/hardware/polarfire/base-packages.nix
Comment (codehub8) [reply modules/hardware/polarfire/base-packages.nix:1]: Tried to use development modules, for package management need to discuss.

File: modules/hardware/polarfire/configuration.nix
Comment (codehub8) [reply modules/hardware/polarfire/configuration.nix]: Created a new varian of networking.nix for such platforms which doesn't use vm.

File: modules/hardware/polarfire/configuration.nix
Comment (codehub8) [reply modules/hardware/polarfire/configuration.nix]: File deleted.

File: modules/hardware/polarfire/configuration.nix
Comment (codehub8) [reply modules/hardware/polarfire/configuration.nix]: We can use docker from development as of now we have decided to disable it by default.

File: modules/hardware/polarfire/configuration.nix
Comment (codehub8) [reply modules/hardware/polarfire/configuration.nix]: File deleted.

File: modules/hardware/polarfire/configuration.nix
Comment (codehub8) [reply modules/hardware/polarfire/configuration.nix]: File deleted.

File: modules/hardware/polarfire/configuration.nix
Comment (codehub8) [reply modules/hardware/polarfire/configuration.nix]: Done.

File: templates/targets/riscv64/microchip/polarfire/flake.nix
Comment (codehub8) [reply templates/targets/riscv64/microchip/polarfire/flake.nix:1]: We decided to integrate it as reference platform.

File: modules/hardware/polarfire/base-packages.nix
Comment (vilvo) [reply modules/hardware/polarfire/base-packages.nix:1]: @brianmcgillion and @mikatammi are in the process of refactoring the ghaf modules behind external API that is configurable instead of static as it has been. They merged a few modules just yesterday. See https://github.com/tiiuae/ghaf/commit/f9e49ec2ded512c0036f77ff4860a59f15fe662f#diff-802221ea9e0a54520f90631da97263b93204b5dec385457827d48e34b001286bR16 for example.

File: modules/hardware/polarfire/configuration.nix
Comment (vilvo) [reply modules/hardware/polarfire/configuration.nix]: docker was made configurable yesterday - disabled by default: https://github.com/tiiuae/ghaf/commit/557a5049aedec2a4053a6717697f170356a410cd
enabling will also require user specific groups configuration.

File: docs/src/features/features.md
Comment (vilvo) [reply docs/src/features/features.md:36]: We still have to address the ghaf `23.05`-only support issue.

File: flake.nix
Comment (vilvo) [inline flake.nix:21]: In particular, we can't merge this as it will break `nixpkgs` via `inputs` for all other targets by taking us back to `22.11` when we have just put the effort to move to `23.05`.

Comment (vilvo) [reply flake.nix:21]: One possible solution is to introduce another input - like `nixpkgs-2211` and pass that to icicle targets instead of `nixpkgs`. The same with `nixos-hardware` below where your fork would be additional input - at least until your upstream PR is merged.
There are some downside from that, namely duplicating the inputs on some areas, but that proposal will allow us to manage this as reference configuration. @mikatammi - please have a look at the proposed changes to `flake.nix` and comment if you can think of other solutions.

Comment (codehub8) [reply flake.nix]: Pushed a new amendment with two more inputs two flake for nixpkgs-22.11 and nixos-hardware forked repository. Now other platforms will not be impacted. Please review.

Comment (vilvo) [reply flake.nix]: Proposed solution has conflicts with the HEAD of main 
screencapture
and in addition it duplicates `nixpkgs` and `nixos-hardware` for all the targets. 
I'm afraid I can't support that solution. Please sync with @mikatammi and @brianmcgillion as they are rewriting the module system which has impact to reviewing / merging any other changes.

**Review by vilvo — CHANGES_REQUESTED** [review]
Duplicating `nixpkgs` and `nixos-hardware` for all the targets is not a good idea.

**Pull request closed without merging**
